### DISCUSSION TITLE
DevOps: What is the best way to create separate Langfuse instances in different environments: Dev, Test, QA, Prod?

### DISCUSSION THREAD

**Discussion by aiakubovich** [langfuse/langfuse#1957] [Support] [answered]
I am wondering about the best way to create instances in different envs.  
  
Let's say we have Dev environments where we have a Langfuse Docker container running alongside a PostgreSQL DB. Now we want to add a new environment, Test.  
  
Should we keep PostgreSQL in Dev and connect both Langfuse instances in Dev and Test to it? Would there be any problems if you have two UI Langfuse interfaces connected to the same DB?  
  
Or should we have separate PostgreSQL instances as well? Would be migration prosses between those if we want to have same data (traces, scores) in both?

**Comment by marcklingen** [comment, marked as answer]
Great question, I'd suggest the following options

- Maintain a single langfuse instance / cluster used across services. Split envs into projects for separated rbac controls on each of the environments in Langfuse.
- Create a fully separate langfuse instance for each of the environments.

I'd discourage to run multiple Langfuse instances in separated environments on the same postgres database and schema. While this works (I mean it is just scaling the stateless api horizontally), this has the risk that updates won't work as expected. The langfuse container applies necessary changes to the database on start. When you run these instances in separated environments, I'd assume that upgrading them all at exactly the same time will be difficult.

**Reply by marcklingen** [reply]
FAQ page on this: https://langfuse.com/faq/all/managing-different-environments

**Comment by marcklingen** [comment]
Update on this: if you manage many environments programmatically, this new features is useful to initialize the instances with default credentials: https://langfuse.com/changelog/2024-09-04-headless-initialization-of-self-hosted-deployments
